# LAB | Introduction to Machine Learning

## Introduction
As a data analyst or data scientist, you will find that while using ML algorithms is the fun and glamorous part of your work. Preparing your data sometimes takes up 90% of your time. Therefore, learning to prepare your data properly is one of the most important skills you will need.
## Resources

[Linear Interpolation](https://en.wikipedia.org/wiki/Linear_interpolation)

[Missing Data Imputation](http://www.stat.columbia.edu/~gelman/arm/missing.pdf)

[7 Steps to Data Preparation](https://www.kdnuggets.com/2017/06/7-steps-mastering-data-preparation-python.html)


## Import your libraries

In [1]:
# You code here
import pandas as pd
import seaborn as sns 
import matplotlib.pyplot as plt


# Challenge 1 - Import and Describe the Dataset

- In this challenge we will use the `austin_weather` data. 
- Import it into a data frame called `austin`. 

In [2]:
# Your code here
austin = pd.read_csv('https://raw.githubusercontent.com/Rairocha/Extra-labs/english-lab/Labs/M03-L01-Introduction_ML/austin_weather.csv')
austin

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0     2013-12-21         74        60        45            67           49   
1     2013-12-22         56        48        39            43           36   
2     2013-12-23         58        45        32            31           27   
3     2013-12-24         61        46        31            36           28   
4     2013-12-25         58        50        41            44           40   
...          ...        ...       ...       ...           ...          ...   
1314  2017-07-27        103        89        75            71           67   
1315  2017-07-28        105        91        76            71           64   
1316  2017-07-29        107        92        77            72           64   
1317  2017-07-30        106        93        79            70           68   
1318  2017-07-31         99        88        77            66           61   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0              43                  93                 75                 57   
1              28                  93                 68                 43   
2              23                  76                 52                 27   
3              21                  89                 56                 22   
4              36                  86                 71                 56   
...           ...                 ...                ...                ...   
1314           61                  82                 54                 25   
1315           55                  87                 54                 20   
1316           55                  82                 51                 19   
1317           63                  69                 48                 27   
1318           54                  64                 43                 22   

      ... SeaLevelPressureAvgInches SeaLevelPressureLowInches  \
0     ...                     29.68                     29.59   
1     ...                     30.13                     29.87   
2     ...                     30.49                     30.41   
3     ...                     30.45                      30.3   
4     ...                     30.33                     30.27   
...   ...                       ...                       ...   
1314  ...                     29.97                     29.88   
1315  ...                      29.9                     29.81   
1316  ...                     29.86                     29.79   
1317  ...                     29.91                     29.87   
1318  ...                     29.97                     29.91   

     VisibilityHighMiles VisibilityAvgMiles VisibilityLowMiles WindHighMPH  \
0                     10                  7                  2          20   
1                     10                 10                  5          16   
2                     10                 10                 10           8   
3                     10                 10                  7          12   
4                     10                 10                  7          10   
...                  ...                ...                ...         ...   
1314                  10                 10                 10          12   
1315                  10                 10                 10          14   
1316                  10                 10                 10          12   
1317                  10                 10                 10          13   
1318                  10                 10                 10          12   

     WindAvgMPH WindGustMPH PrecipitationSumInches               Events  
0             4          31                   0.46  Rain , Thunderstorm  
1             6          25                      0                       
2             3          12                      0                       
3             4          20                      0                       
4             2          16           

### Next, describe the dataset you have loaded: 
- Look at the variables and their types

In [3]:
# Your code here
austin.dtypes

Date                          object
TempHighF                      int64
TempAvgF                       int64
TempLowF                       int64
DewPointHighF                 object
DewPointAvgF                  object
DewPointLowF                  object
HumidityHighPercent           object
HumidityAvgPercent            object
HumidityLowPercent            object
SeaLevelPressureHighInches    object
SeaLevelPressureAvgInches     object
SeaLevelPressureLowInches     object
VisibilityHighMiles           object
VisibilityAvgMiles            object
VisibilityLowMiles            object
WindHighMPH                   object
WindAvgMPH                    object
WindGustMPH                   object
PrecipitationSumInches        object
Events                        object
dtype: object

- Examine the descriptive statistics of the numeric variables 

In [8]:
austin.describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

- Look at the first five rows of all variables to evaluate the categorical variables as well

In [9]:
# Your code here
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [7]:
# Your code here
austin['DewPointHighF'].unique()

array(['67', '43', '31', '36', '44', '39', '41', '49', '51', '37', '55',
       '54', '15', '29', '60', '64', '58', '28', '42', '22', '38', '45',
       '19', '59', '62', '40', '17', '53', '34', '48', '52', '61', '63',
       '33', '57', '65', '46', '32', '50', '27', '56', '47', '68', '69',
       '66', '70', '73', '71', '76', '72', '75', '74', '-', '77', '23',
       '35', '30', '79', '78', '80', '26', '18', '25', '13'], dtype=object)

In [ ]:
# Your observation here

- The following is a list of columns misrepresented as `object`. 
- Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [10]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [16]:
# Your code here
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric,errors='coerce')

In [ ]:
 austin[wrong_type_columns].apply(lambda col: pd.to_numeric(col,errors='coerce'))

In [18]:
pd.read_csv('https://raw.githubusercontent.com/Rairocha/Extra-labs/english-lab/Labs/M03-L01-Introduction_ML/austin_weather.csv',
            na_values=['-']).dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches         object
Events                         object
dtype: object

### Check if your code has worked by printing the data types again. 
- You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [19]:
# Your code here
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

### Now that we have fixed the type mismatch, let's address the missing data.
- By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.
- The first step is to examine how many rows contain missing data.
- We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

### In the next cell, identify all rows containing at least one missing value. 
- Assign the dataframes with missing values to a variable called `missing_values`.

In [23]:
# Your code here
austin.isna().sum(axis=1)
austin.isna().any(axis=1)

0       False
1       False
2       False
3       False
4        True
        ...  
1314    False
1315    False
1316    False
1317    False
1318    False
Length: 1319, dtype: bool

In [41]:
missing_values = austin[austin.isna().any(axis=1)]

In [39]:
austin.dropna()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
5     2013-12-26         57        48        39           39.0          36.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
5             33.0                 79.0                63.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
0                   57.0  ...                      29.68   
1                   43.0  ...                      30.13   
2                   27.0  ...                      30.49   
3                   22.0  ...                      30.45   
5                   47.0  ...                      30.40   
...                  ...  ...                        ...   
1314                25.0  ...                      29.97   
1315                20.0  ...                      29.90   
1316                19.0  ...                      29.86   
1317                27.0  ...                      29.91   
1318                22.0  ...                      29.97   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                         29.59                 10.0                 7.0   
1                         29.87                 10.0                10.0   
2                         30.41                 10.0                10.0   
3                         30.30                 10.0                10.0   
5                         30.34                 10.0                 9.0   
...                         ...                  ...                 ...   
1314                      29.88                 10.0                10.0   
1315                      29.81                 10.0                10.0   
1316                      29.79                 10.0                10.0   
1317                      29.87                 10.0                10.0   
1318                      29.91                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
0                    2.0         20.0         4.0         31.0   
1                    5.0         16.0         6.0         25.0   
2                   10.0          8.0         3.0         12.0   
3                    7.0         12.0         4.0         20.0   
5                    7.0         12.0         3.0         17.0   
...                  ...          ...         ...          ...   
1314                10.0         12.0         5.0         21.0   
1315                10.0         14.0         5.0         20.0   
1316                10.0         12.0         4.0         17.0   


- There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:
> - Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.
> - Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.
> - Filling the values using some algorithm. 

- In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

### Next, count the number of rows of `austin` and `missing_values`.

In [42]:
len(missing_values)

136

#### Calculate the ratio of missing rows to total rows

In [43]:
len(missing_values)/len(austin)

0.10310841546626232

- As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

- We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [44]:
# Your code here
austin.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

- As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. 
- What's the number of missing values in this column in ratio to its total number of rows?

In [46]:
austin.isna().sum()['PrecipitationSumInches']/len(austin)

0.09401061410159212

In [47]:
austin.isna().mean()['PrecipitationSumInches']

0.09401061410159212

- Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 
- It doesn't make sense to *guess* its missing values because the estimation will be too 

### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [50]:
austin = austin.drop(columns='PrecipitationSumInches')

In [51]:
# Print `austin` to confirm the column is indeed removed
austin.columns

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'Events'],
      dtype='object')

### Next we will perform linear interpolation of the missing data.

- This means that we will use a linear algorithm to estimate the missing data. 
- Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is preferable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

- If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

- In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [54]:
austin_fixed = austin.interpolate()

C:\Users\raian\AppData\Local\Temp\ipykernel_24500\112129401.py:1: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  austin_fixed = austin.interpolate()


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [55]:
austin_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

# Challenge 3 - Processing the `Events` Column

### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.
- Use the `value_counts()` function to evaluate the contents of this column

In [56]:
austin_fixed['Events'].value_counts()

Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

- Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [59]:
# Your answer
austin_fixed['Events'].str.split(',').str.len().max()

3

### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

- How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

```python
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']
```

In [64]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']
for event in event_list:
    austin_fixed[event]= 0

In [65]:
# Print your new dataframe to check whether new columns have been created:
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
0            20.0         4.0         31.0  Rain , Thunderstorm     0    0   
1            16.0         6.0         25.0                          0    0   
2             8.0         3.0         12.0                          0    0   
3            12.0         4.0         20.0                          0    0   
4            10.0         2.0         16.0                          0    0   
...           ...         ...          ...                  ...   ...  ...   
1314         12.0         5.0         21.0                          0    0   
1315         14.0         5.0         20.0                          0    0   
1316         12.0         4.0         17.0                          0    0   
1317         13.0         4.0         20.0                          0    0   
1318         12.0         4.0         20.0                          0    0   

      Rain Thunderstorm  
0        0            0  
1        0            0  
2        0            0  
3        0            0  
4        0            0  
...    ...          ...  
1314     0            0  
1315     0            0  
1316     0            0  
1317     0            0  
1318     0            0  

[1319 rows x 24 columns]

### Next, populate the actual values in the dummy columns of  `austin_fixed`.

- You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [67]:
# Your code here
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']
for event in event_list:
    austin_fixed[event] = austin_fixed['Events'].str.contains(event).astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [68]:
# Your code here
    
austin_fixed.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0  ...                 7.0                 2.0         20.0         4.0   
1  ...                10.0                 5.0         16.0         6.0   
2  ...                10.0                10.0          8.0         3.0   
3  ...                10.0                 7.0         12.0         4.0   
4  ...                10.0                 7.0         10.0         2.0   

   WindGustMPH               Events  Snow  Fog  Rain Thunderstorm  
0         31.0  Rain , Thunderstorm     0    0     1            1  
1         25.0                          0    0     0            0  
2         12.0                          0    0     0            0  
3         20.0                          0    0     0            0  
4         16.0                          0    0     0            0  

[5 rows x 24 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [70]:
# Your code here
austin_fixed = austin_fixed.drop(columns = 'Events')


In [71]:
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   57.0  ...                 10.0                 7.0   
1                   43.0  ...                 10.0                10.0   
2                   27.0  ...                 10.0                10.0   
3                   22.0  ...                 10.0                10.0   
4                   56.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                25.0  ...                 10.0                10.0   
1315                20.0  ...                 10.0                10.0   
1316                19.0  ...                 10.0                10.0   
1317                27.0  ...                 10.0                10.0   
1318                22.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 23 columns]

In [80]:
# altenative treatment
events_table  = austin['Events'].str.split(', ').explode().str.strip()
events_table[events_table != '']

0               Rain
0       Thunderstorm
18              Rain
19               Fog
20              Rain
            ...     
1304    Thunderstorm
1310            Rain
1310    Thunderstorm
1311            Rain
1311    Thunderstorm
Name: Events, Length: 635, dtype: object

# Challenge 4 - Processing The `Date` Column
- The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. 
- One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. 
- In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [86]:
from datetime import datetime
datetime.toordinal(datetime.strptime('2014-01-06','%Y-%m-%d'))

735239

In [91]:
datetime.strptime('2014-01-06','%Y-%m-%d').toordinal()

735239

In [97]:
austin_fixed['Date'].apply(lambda d : datetime.toordinal(datetime.strptime(d,'%Y-%m-%d')))
austin_fixed['Date'].apply(lambda d : datetime.strptime(d,'%Y-%m-%d').toordinal())
austin_fixed['Date'] = pd.to_datetime(austin_fixed['Date']).apply(datetime.toordinal)

0       735223
1       735224
2       735225
3       735226
4       735227
         ...  
1314    736537
1315    736538
1316    736539
1317    736540
1318    736541
Name: Date, Length: 1319, dtype: int64

#### Print `austin_fixed` to check your `Date` column.

In [98]:
# Your code here
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     735223         74        60        45           67.0          49.0   
1     735224         56        48        39           43.0          36.0   
2     735225         58        45        32           31.0          27.0   
3     735226         61        46        31           36.0          28.0   
4     735227         58        50        41           44.0          40.0   
...      ...        ...       ...       ...            ...           ...   
1314  736537        103        89        75           71.0          67.0   
1315  736538        105        91        76           71.0          64.0   
1316  736539        107        92        77           72.0          64.0   
1317  736540        106        93        79           70.0          68.0   
1318  736541         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   57.0  ...                 10.0                 7.0   
1                   43.0  ...                 10.0                10.0   
2                   27.0  ...                 10.0                10.0   
3                   22.0  ...                 10.0                10.0   
4                   56.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                25.0  ...                 10.0                10.0   
1315                20.0  ...                 10.0                10.0   
1316                19.0  ...                 10.0                10.0   
1317                27.0  ...                 10.0                10.0   
1318                22.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

### Now that we have processed the data for machine learning, we will separate the data to test and training sets.
- We first train the model using only the training set. 
- We check our metrics on the training set. 
- We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. 
- We will need to revise our model to ensure it will be more applicable to data outside the test set.

### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

- When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [101]:
# Your code here
X = austin_fixed.drop(columns='TempAvgF')
X = austin_fixed[['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]
y = austin_fixed['TempAvgF']

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [134]:
# Your code here
from sklearn.model_selection import train_test_split
X_train, X_test, y_train,y_test= train_test_split(X,y,train_size=0.8,shuffle=False)

In [135]:
X_train

Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0     735223         74        45           67.0          49.0          43.0   
1     735224         56        39           43.0          36.0          28.0   
2     735225         58        32           31.0          27.0          23.0   
3     735226         61        31           36.0          28.0          21.0   
4     735227         58        41           44.0          40.0          36.0   
...      ...        ...       ...            ...           ...           ...   
1050  736273         72        63           68.0          66.0          59.0   
1051  736274         76        59           67.0          60.0          57.0   
1052  736275         71        61           69.0          65.0          61.0   
1053  736276         69        62           66.0          64.0          62.0   
1054  736277         65        60           62.0          61.0          58.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0                    93.0                75.0                57.0   
1                    93.0                68.0                43.0   
2                    76.0                52.0                27.0   
3                    89.0                56.0                22.0   
4                    86.0                71.0                56.0   
...                   ...                 ...                 ...   
1050                100.0                91.0                81.0   
1051                 93.0                85.0                76.0   
1052                100.0                94.0                87.0   
1053                100.0                92.0                84.0   
1054                100.0                94.0                87.0   

      SeaLevelPressureHighInches  ...  VisibilityHighMiles  \
0                          29.86  ...                 10.0   
1                          30.41  ...                 10.0   
2                          30.56  ...                 10.0   
3                          30.56  ...                 10.0   
4                          30.41  ...                 10.0   
...                          ...  ...                  ...   
1050                       30.27  ...                 10.0   
1051                       30.19  ...                 10.0   
1052                       30.14  ...                 10.0   
1053                       30.21  ...                 10.0   
1054                       30.29  ...                 10.0   

      VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                    7.0                 2.0         20.0         4.0   
1                   10.0                 5.0         16.0         6.0   
2                   10.0                10.0          8.0         3.0   
3                   10.0                 7.0         12.0         4.0   
4                   10.0                 7.0         10.0         2.0   
...                  ...                 ...          ...         ...   
1050                 6.0                 1.0          7.0         2.0   
1051                10.0                 5.0         12.0         2.0   
1052                 5.0                 1.0          7.0         2.0   
1053                 8.0                 2.0         10.0         4.0   
1054                 6.0                 1.0         10.0         6.0   

      WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0            31.0     0    0     1             1  
1            25.0     0    0     0             0  
2            12.0     0    0     0             0  
3            20.0     0    0     0             0  
4            16.0     0    0     0             0  
...           ...   ...  ...   ...           ...  
1050         12.0     0    0     1             0  
1051         20.0     0    0     1             0  
1052         12.0     0    0     1             0  
1053         16.0     0    0     1             1  
1054         17.0     0    0     1 

- Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [ ]:
#Your code here

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

- This is because many time series algorithms rely on observations having equal time distances between them. 
- In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. 
- We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here